File: Grind75/38-MinStack/cpp/MinStack.cpp
#include "MinStack.h"
#include <stack>
#include <utility>

using namespace std;

void MinStack::push(int val) {
  if (stack.empty())
    stack.push(make_pair(val, val));
  else
    stack.push(make_pair(val, min(getMin(), val)));
}

void MinStack::pop() {
  stack.pop();
}

int MinStack::top() {
  return stack.top().first;
}

int MinStack::getMin() {
  return stack.top().second;
}


void MinStack2::push(int val) {
  Node *newTop;
  if (!topNode)
    newTop = new Node(val, val, nullptr);
  else
    newTop = new Node(val, min(topNode->min, val), topNode);
  topNode = newTop;
}

void MinStack2::pop() {
  Node *toPop = topNode;
  topNode = topNode->below;
  delete toPop;
}

int MinStack2::top() {
  return topNode->value;
}

int MinStack2::getMin() {
  return topNode->min;
}

File: Grind75/38-MinStack/python/MinStack.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')
from collections import deque

class MinStack:
  def __init__(self):
    self._valstack = deque()
    self._minstack = deque()

  def push(self, val: int) -> None:
    self._valstack.append(val)
    if len(self._minstack) > 0:
      self._minstack.append(min(self.getMin(),val))
    else:
      self._minstack.append(val)

  def pop(self) -> None:
    self._valstack.pop()
    self._minstack.pop()

  def top(self) -> int:
    return self._valstack[-1]

  def getMin(self) -> int:
    return self._minstack[-1]

if __name__ == '__main__':
  ms = MinStack()
  ms.push(2)
  assert(ms.top()    == 2)
  assert(ms.getMin() == 2)
  ms.push(1)
  assert(ms.top()    == 1)
  assert(ms.getMin() == 1)
  ms.push(3)
  assert(ms.top()    == 3)
  assert(ms.getMin() == 1)


